core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1872 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use self::{
+    HandleType::{
+        Internal,
+        Leaf,
+    },
+    InsertResult::{
+        Fit,
+        Split,
+    },
+};
+use super::search::{
+    self,
+    SearchResult::{
+        Found,
+        NotFound,
+    },
+};
+use crate::storage::{
+    self,
+    alloc::{
+        Allocate,
+        AllocateUsing,
+        Initialize,
+    },
+    chunk::SyncChunk,
+    Flush,
+};
+use core::{
+    borrow::Borrow,
+    cmp::Ord,
+    ptr,
+};
+#[cfg(feature = "ink-generate-abi")]
+use ink_abi::{
+    HasLayout,
+    LayoutField,
+    LayoutStruct,
+    StorageLayout,
+};
+use scale::{
+    Codec,
+    Decode,
+    Encode,
+};
+#[cfg(feature = "ink-generate-abi")]
+use type_metadata::Metadata;
+
+/// Each node in the tree has 2 * B children.
```

Comment:
I chose the number solely based on the reason that the Rust standard library implementation [uses the same number](https://github.com/rust-lang/rust/blob/master/src/liballoc/collections/btree/node.rs#L42). Since we don't have empirical data yet I took this number as a first starting point.

But yeah, I agree, there needs to be a logical argument for the number of children. Once people use the data structure it will be hard to migrate to a new number.

Reply:
So from what I understand (e.g. from [this](http://cglab.ca/~abeinges/blah/rust-btree-case/) extensive description of an earlier implementation) the `6` also doesn't have that much of a reasoning behind it, but rather seemed like a good value for the linear search conducted within a node when searching for a key while also keeping the algorithm cache-friendly.

I think for our case the best thing would be to have some proper benchmarks. Though I can suppose the results will vary widely depending on what you put into the `BTreeMap`.